File: app/src/test/java/com/beautifulquran/domain/EnglishTypographyTest.kt
package com.beautifulquran.domain

import org.junit.Assert.assertEquals
import org.junit.Test

class EnglishTypographyTest {
    @Test
    fun `adds a stop only at the ayah end`() {
        assertEquals(
            listOf("and killed", "Dawood", "Jalut", "And if not."),
            EnglishTypography.punctuate(
                listOf("and killed", "Dawood", "Jalut", "And if not"),
            ),
        )
    }

    @Test
    fun `does not infer internal sentence boundaries from capitals`() {
        assertEquals(
            listOf("that which", "He willed", "And if not", "the worlds."),
            EnglishTypography.punctuate(
                listOf("that which", "He willed", "And if not", "the worlds"),
            ),
        )
    }

    @Test
    fun `preserves existing punctuation`() {
        assertEquals(
            listOf("Why?", "Then", "listen!"),
            EnglishTypography.punctuate(
                listOf("Why?", "Then", "listen!"),
            ),
        )
    }

    @Test
    fun `coalesces a shared phrase spanning different Arabic words`() {
        assertEquals(
            listOf("guide", "the wrongdoing people.", ""),
            EnglishTypography.lyricize(
                glosses = listOf("guide", "the wrongdoing people", "the wrongdoing people"),
                arabicWords = listOf(
                    "يَهۡدِي",
                    "ٱلۡقَوۡمَ",
                    "ٱلظَّـٰلِمِينَ",
                ),
            ),
        )
    }

    @Test
    fun `keeps a genuine repeated Arabic word`() {
        assertEquals(
            listOf("a saying", "Peace", "Peace."),
            EnglishTypography.lyricize(
                glosses = listOf("a saying", "Peace", "Peace"),
                arabicWords = listOf("قِيلٰا", "سَلَـٰمٰا", "سَلَـٰمٰا"),
            ),
        )
    }
}

File: web/src/domain/__tests__/EnglishTypography.test.ts
import { describe, expect, it } from 'vitest'
import {
  coalescedGlossOwnerIndex,
  lyricizeEnglishGlosses,
  punctuateEnglishGlosses,
} from '../EnglishTypography'

describe('punctuateEnglishGlosses', () => {
  it('adds a stop only at the ayah end', () => {
    expect(punctuateEnglishGlosses(['and killed', 'Dawood', 'Jalut', 'And if not']))
      .toEqual(['and killed', 'Dawood', 'Jalut', 'And if not.'])
  })

  it('does not infer internal sentence boundaries from capitals', () => {
    expect(punctuateEnglishGlosses(['that which', 'He willed', 'And if not', 'the worlds']))
      .toEqual(['that which', 'He willed', 'And if not', 'the worlds.'])
  })

  it('preserves existing punctuation', () => {
    expect(punctuateEnglishGlosses(['Why?', 'Then', 'listen!']))
      .toEqual(['Why?', 'Then', 'listen!'])
  })

  it('coalesces a shared phrase spanning different Arabic words', () => {
    expect(lyricizeEnglishGlosses(
      ['guide', 'the wrongdoing people', 'the wrongdoing people'],
      ['يَهۡدِي', 'ٱلۡقَوۡمَ', 'ٱلظَّـٰلِمِينَ'],
    )).toEqual(['guide', 'the wrongdoing people.', ''])
  })

  it('keeps a genuine repeated Arabic word', () => {
    expect(lyricizeEnglishGlosses(
      ['a saying', 'Peace', 'Peace'],
      ['قِيلٰا', 'سَلَـٰمٰا', 'سَلَـٰمٰا'],
    )).toEqual(['a saying', 'Peace', 'Peace.'])
  })

  it('resolves a shared-gloss flash to its visible owner', () => {
    const glosses = ['Except', 'righteous deeds', 'righteous deeds', 'then']
    const arabic = ['إِلَّا', 'وَعَمِلَ', 'صَٰلِحٗا', 'فَأُوْلَٰٓئِكَ']
    expect(coalescedGlossOwnerIndex(glosses, arabic, 2)).toBe(1)
    expect(coalescedGlossOwnerIndex(glosses, arabic, 1)).toBe(1)
    expect(coalescedGlossOwnerIndex(glosses, arabic, -1)).toBeNull()
  })
})
